README
NAME
    DBIx::Class::InflateColumn::Object::Enum - Allows a DBIx::Class user to
    define a Object::Enum column

VERSION
    Version 0.03

SYNOPSIS
    Load this module via load_components and utilize is_enum and values
    property to define Enumuration columns via Object::Enum

        package TableClass;
    
        use strict;
        use warnings;
        use base 'DBIx::Class';
    
        __PACKAGE__->load_components(qw/InflateColumn::Object::Enum Core/);
        __PACKAGE__->table('testtable');
        __PACKAGE__->add_columns(
            color => {
                data_type => 'varchar',
                is_enum => 1,
                extra => {
                    list => [qw/red green blue/]
                }
            }
            color_native => { # works inline with native enum type
                data_type => 'enum',
                is_enum => 1,
                extra => {
                    list => [qw/red green blue/]
                }
            }
        );
    
        1;
    
    Now you may treat the column as an Object::Enum object.

        my $table_rs = $db->resultset('TableClass')->create({
            color => undef
        });
    
        $table_rs->color->set_red; # sets color to red
        $table_rs->color->is_red; # would return true
        $table_rs->color->is_green; # would return false
        print $table_rs->color->value; # would print 'red'
        $table_rs->color->unset; # set the value to 'undef' or 'null'
        $table_rs->color->is_red; # returns false now
    
METHODS
  register_column
    Internal chained method with "register_column" in DBIx::Class::Row.
    Users do not call this directly!

AUTHOR
    Jason M. Mills, "<jmmills at cpan.org>"

BUGS
    Please report any bugs or feature requests to
    "bug-dbix-class-inflatecolumn-object-enum at rt.cpan.org", or through
    the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=DBIx-Class-InflateColumn
    -Object-Enum>. I will be notified, and then you'll automatically be
    notified of progress on your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc DBIx::Class::InflateColumn::Object::Enum

    You can also look for information at:

    * RT: CPAN's request tracker
        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=DBIx-Class-InflateColumn-O
        bject-Enum>

    * AnnoCPAN: Annotated CPAN documentation
        <http://annocpan.org/dist/DBIx-Class-InflateColumn-Object-Enum>

    * CPAN Ratings
        <http://cpanratings.perl.org/d/DBIx-Class-InflateColumn-Object-Enum>

    * Search CPAN
        <http://search.cpan.org/dist/DBIx-Class-InflateColumn-Object-Enum>

SEE ALSO
    Object::Enum, DBIx::Class, DBIx::Class::InflateColumn::URI

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Foo

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Foo

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Foo

    CPAN Ratings
        http://cpanratings.perl.org/d/Foo

    Search CPAN
        http://search.cpan.org/dist/Foo


COPYRIGHT AND LICENCE

Copyright (C) 2009 Jason M. Mills

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



Changes
Revision history for DBIx-Class-InflateColumn-Object-Enum

0.05 Jan 27th 2014
	- Updated dist build to distzilla
	- fixed https://rt.cpan.org/Ticket/Display.html?id=93432
	- fixed https://rt.cpan.org/Ticket/Display.html?id=74086
	* TODO: fix issue 80705
0.04    Jan 27th 2009
	- Removed self dependency 
0.03    Nov 11th 2008
	- Modified error verbage as per request from mst
	- Changed ChangeLog sorting from latest -> oldest
	- Migrated from internal svn repo to dbic-class svn repo

0.02    Oct 25th 2008
        - Expanded testing database and tap tests to test for varchar 
          nullable and non-nullable and database native enum nullable 
          and non-nullable 
        - Fixed a implmentation for Object::Enum->new so that columns value
          on inflation is set correctly. This could have been considered
          a nasty bug. Discovered through new tests.
        - Added sanity check for inflation value set for invalid values so 
          that we don't get Object::Enum's die for invalid value as I don't
          like that behavior. Maybe a column_info to control check would be 
          more reasonable?
        - Changed allowable values usages from values column_info property
          to extra->list property to make this modules usage fall more 
          inline with how SQL::Translator handles native enum database
          types. As per suggestion from Bernhard Graf.
	- Added to code to enable smarter default_value handling

0.01    Oct 23rd 2008
        - Register_column implemented
        - Test database schema implemented
        - Wrote base TAP tests


t/lib/TestDB/Test.pm
package TestDB::Test;

use strict;
use warnings;

use base 'DBIx::Class';

__PACKAGE__->load_components(qw/
    InflateColumn::Object::Enum
    PK::Auto
    Core
/);
__PACKAGE__->table('test');
__PACKAGE__->add_columns(
    id => {
        data_type => 'number',
        is_auto_increment => 1,
        is_nullable => 0
    },
    name => {
        data_type => 'varchar',
        size => '64',
        is_nullable => 0
    },
    color => {
        data_type => 'varchar',
        is_nullable => 1,
        size => '64',
        is_enum => 1,
        values => [qw/red green blue]/]
    }
);
__PACKAGE__->set_primary_key('id');

1;

t/lib/TestDB/NativeEnumNullable.pm
package TestDB::NativeEnumNullable;

use strict;
use warnings;

use base 'DBIx::Class';

__PACKAGE__->load_components(qw/
    InflateColumn::Object::Enum
    PK::Auto
    Core
/);
__PACKAGE__->table('nen');
__PACKAGE__->add_columns(
    id => {
        data_type => 'number',
        is_auto_increment => 1,
        is_nullable => 0
    },
    enum => {
        data_type => 'enum',
        is_enum => 1,
        is_nullable => 1,
        extra => {
            list => [qw/red green blue/]
        },
    }
);
__PACKAGE__->set_primary_key('id');

1;


t/lib/TestDB/VarcharEnumNullable.pm
package TestDB::VarcharEnumNullable;

use strict;
use warnings;

use base 'DBIx::Class';

__PACKAGE__->load_components(qw/
    InflateColumn::Object::Enum
    PK::Auto
    Core
/);
__PACKAGE__->table('ven');
__PACKAGE__->add_columns(
    id => {
        data_type => 'number',
        is_auto_increment => 1,
        is_nullable => 0
    },
    enum => {
        data_type => 'varchar',
        is_enum => 1,
        is_nullable => 1,
        extra => {
            list => [qw/red green blue/]
        },
    }
);
__PACKAGE__->set_primary_key('id');

1;


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.013.
use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Allows a DBIx::Class user to define a Object::Enum column",
  "AUTHOR" => "Jason M. Mills <jmmills\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "DBIx-Class-InflateColumn-Object-Enum",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "DBIx::Class::InflateColumn::Object::Enum",
  "PREREQ_PM" => {
    "DBIx::Class::Schema" => 0,
    "Object::Enum" => 0
  },
  "TEST_REQUIRES" => {
    "DBICx::TestDatabase" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.05",
  "test" => {
    "TESTS" => "t/*.t t/regres/*.t"
  }
);


my %FallbackPrereqs = (
  "DBICx::TestDatabase" => 0,
  "DBIx::Class::Schema" => 0,
  "Object::Enum" => 0,
  "Test::More" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/DBIx/Class/InflateColumn/Object/Enum.pm
package DBIx::Class::InflateColumn::Object::Enum;
$DBIx::Class::InflateColumn::Object::Enum::VERSION = '0.05';
use warnings;
use strict;
use Carp qw/croak confess/;
use Object::Enum;

=head1 NAME

DBIx::Class::InflateColumn::Object::Enum - Allows a DBIx::Class user to define a Object::Enum column

=head1 VERSION

Version 0.03

=cut

# Dist::Zill should handle this now
# ABSTRACT:  Allows a DBIx::Class user to define a Object::Enum column
#our $VERSION = '0.04';


=head1 SYNOPSIS

Load this module via load_components and utilize is_enum and values property
to define Enumuration columns via Object::Enum

    package TableClass;
    
    use strict;
    use warnings;
    use base 'DBIx::Class';
    
    __PACKAGE__->load_components(qw/InflateColumn::Object::Enum Core/);
    __PACKAGE__->table('testtable');
    __PACKAGE__->add_columns(
        color => {
            data_type => 'varchar',
            is_enum => 1,
            extra => {
                list => [qw/red green blue/]
            }
        }
        color_native => { # works inline with native enum type
            data_type => 'enum',
            is_enum => 1,
            extra => {
                list => [qw/red green blue/]
            }
        }
    );
    
    1;
    
Now you may treat the column as an L<Object::Enum> object.
    
    my $table_rs = $db->resultset('TableClass')->create({
        color => undef
    });
    
    $table_rs->color->set_red; # sets color to red
    $table_rs->color->is_red; # would return true
    $table_rs->color->is_green; # would return false
    print $table_rs->color->value; # would print 'red'
    $table_rs->color->unset; # set the value to 'undef' or 'null'
    $table_rs->color->is_red; # returns false now
    

=head1 METHODS

=head2 register_column

Internal chained method with L<DBIx::Class::Row/register_column>.
Users do not call this directly!

=cut

sub register_column {
    my $self = shift;
    my ($column, $info) = @_;
    
    $self->next::method(@_);
    
    return unless defined $info->{is_enum} and $info->{is_enum};
    
    croak("Object::Enum '$column' missing 'extra => { list => [] }' column configuration")
        unless (
            defined $info->{extra}
            and ref $info->{extra}  eq 'HASH'
            and defined $info->{extra}->{list}
        );
        
    croak("Object::Enum '$column' value list (extra => { list => [] }) must be an ARRAY reference")
        unless ref $info->{extra}->{list} eq 'ARRAY';
    
    my $values = $info->{extra}->{list};
    my %values = map {$_=>1} @{$values};

    push(@{$values},$info->{default_value})
        if defined($info->{default_value})
        && !exists $values{$info->{default_value}};

    push(@{$values}, undef)
        if defined($info->{is_nullable})
        && $info->{is_nullable};

    $self->inflate_column(
        $column => {
            inflate => sub {
                my $val = shift;
                my $e = Object::Enum->new({values=>$values});
                $e->value($val);
                return $e;
            },
            deflate => sub {
                return shift->value
            }
        }
    );
    
}

=head1 AUTHOR

Jason M. Mills, C<< <jmmills at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-dbix-class-inflatecolumn-object-enum at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=DBIx-Class-InflateColumn-Object-Enum>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc DBIx::Class::InflateColumn::Object::Enum


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=DBIx-Class-InflateColumn-Object-Enum>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/DBIx-Class-InflateColumn-Object-Enum>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/DBIx-Class-InflateColumn-Object-Enum>

=item * Search CPAN

L<http://search.cpan.org/dist/DBIx-Class-InflateColumn-Object-Enum>

=back


=head1 SEE ALSO

L<Object::Enum>, L<DBIx::Class>, L<DBIx::Class::InflateColumn::URI>


=head1 COPYRIGHT & LICENSE

Copyright 2008 Jason M. Mills, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


=cut

1; # End of DBIx::Class::InflateColumn::Object::Enum
